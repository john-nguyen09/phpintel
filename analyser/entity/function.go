package entity

import (
	"github.com/sourcegraph/go-lsp"
)

type Function struct {
	Location lsp.Range
	Name     string
	Type     string
	Params   []Param
}
