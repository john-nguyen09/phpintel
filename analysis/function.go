package analysis

import (
	"github.com/john-nguyen09/go-phpparser/lexer"
	"github.com/john-nguyen09/go-phpparser/phrase"
	"github.com/john-nguyen09/phpintel/util"
	"github.com/sourcegraph/go-lsp"
)

// Function contains information of functions
type Function struct {
	location lsp.Location
	document *Document

	Name   string `json:"Name"`
	Params []Parameter
}

func newFunction(document *Document, node *phrase.Phrase) Symbol {
	function := &Function{
		location: document.GetNodeLocation(node),
		document: document,
		Params:   make([]Parameter, 0),
	}
	document.pushVariableTable()

	traverser := util.NewTraverser(node)
	child := traverser.Advance()
	for child != nil {
		if p, ok := util.IsOfPhraseTypes(child, []phrase.PhraseType{
			phrase.FunctionDeclarationHeader,
			phrase.MethodDeclarationHeader,
		}); ok {
			function.analyseHeader(p)
		}
		if p, ok := util.IsOfPhraseTypes(child, []phrase.PhraseType{
			phrase.FunctionDeclarationBody,
			phrase.MethodDeclarationBody,
		}); ok {
			scanForChildren(document, p)
		}
		child = traverser.Advance()
	}

	return function
}

func (s *Function) analyseHeader(node *phrase.Phrase) {
	traverser := util.NewTraverser(node)
	child := traverser.Advance()
	for child != nil {
		if token, ok := child.(*lexer.Token); ok {
			switch token.Type {
			case lexer.Name:
				{
					s.Name = util.GetNodeText(token, s.document.text)
				}
			}
		} else if p, ok := child.(*phrase.Phrase); ok {
			switch p.Type {
			case phrase.ParameterDeclarationList:
				{
					s.analyseParameterDeclarationList(p)
				}
			case phrase.Identifier:
				s.Name = util.GetNodeText(p, s.document.text)
			}
		}
		child = traverser.Advance()
	}
}

func (s *Function) analyseParameterDeclarationList(node *phrase.Phrase) {
	traverser := util.NewTraverser(node)
	child := traverser.Advance()
	for child != nil {
		if p, ok := child.(*phrase.Phrase); ok && p.Type == phrase.ParameterDeclaration {
			param := newParameter(s.document, p)
			s.Params = append(s.Params, *param)
		}

		child = traverser.Advance()
	}
}

func (s *Function) getLocation() lsp.Location {
	return s.location
}

func (s *Function) getDocument() *Document {
	return s.document
}
