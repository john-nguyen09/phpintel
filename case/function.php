<?php

function TestFunc1($param1, $param2, $param3 = true, $param4 = '') {

}

function TestFunc2($param1 = null, int $param2) {
    return 1;
}

function TestFunc3($param1, $param2, $param3, bool $param4) {
    if ($param1) {
        return true;
    }

    if ($param2) {
        return 3.14;
    }

    if ($param3) {
        return 15;
    }

    if (!$param4) {
        return null;
    }

    return $param4;
}
