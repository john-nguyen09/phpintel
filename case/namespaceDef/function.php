<?php
namespace NamespaceDef;

/**
 * @param TestClassNamespace1 $param1
 * @param \FullyQualified\Name\Class $param2
 * @param boolean
 * 
 * @return TestClassNamespace2
 */
function function_with_namespace($param1, $param2, bool $param3) {
    $var1 = false;
    $var2 = 'abc';
    $var3 = CONSTANT_IN_NAMESPACE;

    if ($var1 && $var2 && $var3) {
        return $var3;
    }

    if ($var1 && $var2) {
        return $var2;
    }

    if ($var1) {
        return $param1;
    }

    if ($var2) {
        return $param2;
    }

    if ($var3) {
        return $param3;
    }
}