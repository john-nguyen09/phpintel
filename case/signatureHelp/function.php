<?php

test_function1();
