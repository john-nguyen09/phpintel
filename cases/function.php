<?php

function testFunction(TestClass $param1 = null) {
    
}

/**
 * function3's description
 * 
 * @return TestClass
 */
function function3() {

}
