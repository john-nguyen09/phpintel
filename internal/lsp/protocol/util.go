package protocol

import (
	"strconv"
	"strings"
)

// IsInRange compare whether a position is within a range
func IsInRange(position Position, theRange Range) int {
	start := theRange.Start
	end := theRange.End

	if position.Line < start.Line || (position.Line == start.Line && position.Character < start.Character) {
		return -1
	}
	if position.Line > end.Line || (position.Line == end.Line && position.Character >= end.Character) {
		if ComparePos(start, end) == 0 {
			return -1
		}
		return 1
	}
	return 0
}

// ComparePos returns -1, 0 or 1 indicating whether a is before, equal to or after b
func ComparePos(a Position, b Position) int {
	if a.Line < b.Line {
		return -1
	}
	if a.Line == b.Line {
		if a.Character < b.Character {
			return -1
		}
		if a.Character == b.Character {
			return 0
		}
		if a.Character > b.Character {
			return 1
		}
	}
	return 1
}

// CompareRange compares 2 ranges
func CompareRange(a Range, b Range) int {
	if IsInRange(a.End, b) < 0 {
		return -1
	}
	if IsInRange(a.Start, b) > 0 {
		return 1
	}
	return 0
}

// PosFromString creates Position from string
func PosFromString(str string) Position {
	values := strings.Split(str, ":")
	line, err := strconv.Atoi(values[0])
	if err != nil {
		panic(err)
	}
	character, err := strconv.Atoi(values[1])
	if err != nil {
		panic(err)
	}
	return Position{
		Line:      line,
		Character: character,
	}
}

// RangeFromString creates Range from string
func RangeFromString(str string) Range {
	values := strings.Split(str, "-")
	start := PosFromString(values[0])
	end := PosFromString(values[1])
	return Range{
		Start: start,
		End:   end,
	}
}
