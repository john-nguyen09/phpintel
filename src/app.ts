import "reflect-metadata";
import { LevelDatasource } from "./storage/db";
import { LogWriter } from "./service/logWriter";
import { Hasher } from "./service/hasher";
import { Container, interfaces } from "inversify";
import { Indexer } from "./index/indexer";
import { ClassTable } from "./storage/table/class";
import { ClassConstantTable } from "./storage/table/classConstant";
import { ConstantTable } from "./storage/table/constant";
import { FunctionTable } from "./storage/table/function";
import { MethodTable } from "./storage/table/method";
import { PropertyTable } from "./storage/table/property";
import { Traverser } from "./traverser";
import { ReferenceTable } from "./storage/table/reference";
import { PhpDocumentTable } from "./storage/table/phpDoc";
import { IConnection } from "vscode-languageserver";
import { ScopeVarTable } from "./storage/table/scopeVar";

export interface AppOptions {
    storage: string;
}

export class Application {
    protected container: Container = new Container();
    protected options: AppOptions = {
        storage: ''
    };

    constructor(storageLocation: string, connection?: IConnection) {
        this.container.bind<IConnection | undefined>('IConnection').toConstantValue(connection);

        this.initBind();
        this.initStorage(storageLocation);
    }

    public getContainer(): Container {
        return this.container;
    }

    public async clearCache() {
        const db = this.container.get<LevelDatasource>(LevelDatasource).getDb();
        let promises: Promise<void>[] = [];

        return new Promise<void>((resolve, reject) => {
            db.createKeyStream()
                .on('data', (key) => {
                    promises.push(db.del(key));
                })
                .on('end', () => {
                    Promise.all(promises).then(() => { resolve() });
                })
                .on('error', (err: Error) => {
                    reject(err);
                });
        });
    }

    public async shutdown() {
        await this.container.get<LevelDatasource>(LevelDatasource).getDb().close();
    }

    protected initStorage(location: string) {
        this.options.storage = location;

        let datasource = new LevelDatasource(location);
        this.container.bind<LevelDatasource>(LevelDatasource).toConstantValue(datasource);

        // Tables
        this.container.bind<ClassTable>(ClassTable).toSelf().inSingletonScope();
        this.container.bind<ClassConstantTable>(ClassConstantTable).toSelf().inSingletonScope();
        this.container.bind<ConstantTable>(ConstantTable).toSelf().inSingletonScope();
        this.container.bind<FunctionTable>(FunctionTable).toSelf().inSingletonScope();
        this.container.bind<MethodTable>(MethodTable).toSelf().inSingletonScope();
        this.container.bind<PropertyTable>(PropertyTable).toSelf().inSingletonScope();
        this.container.bind<PhpDocumentTable>(PhpDocumentTable).toSelf().inSingletonScope();
        this.container.bind<ReferenceTable>(ReferenceTable).toSelf().inSingletonScope();
        this.container.bind<ScopeVarTable>(ScopeVarTable).toSelf().inSingletonScope();
    }

    protected initBind() {
        this.container.bind<LogWriter>(LogWriter).toSelf().inSingletonScope();
        this.container.bind<Hasher>(Hasher).toSelf().inSingletonScope();
        this.container.bind<Indexer>(Indexer).toSelf().inSingletonScope();
        this.container.bind<Traverser>(Traverser).toSelf().inSingletonScope();
    }
}

export namespace App {
    let app: Application;

    export function init(storageLocation: string, connection?: IConnection) {
        app = new Application(storageLocation, connection);
    }

    export function get<T>(
        identifier: string | symbol | interfaces.Newable<T> | interfaces.Abstract<T>
    ): T {
        return app.getContainer().get(identifier);
    }

    export function set<T>(
        identifier: string,
        value: T
    ): void {
        app.getContainer().bind<T>(identifier).toConstantValue(value);
    }

    export async function clearCache() {
        return await app.clearCache();
    }

    export async function shutdown() {
        return await app.shutdown();
    }
}
