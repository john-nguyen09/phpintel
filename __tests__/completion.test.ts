import { App } from "../src/app";
import * as path from "path";
import { getDebugDir, getCaseDir } from "../src/testHelper";
import { FunctionTable } from "../src/storage/table/function";
import { Indexer, PhpFileInfo } from "../src/index/indexer";
import { PhpDocumentTable } from "../src/storage/table/phpDoc";
import { pathToUri } from "../src/util/uri";
import { ReferenceTable } from "../src/storage/table/referenceTable";
import { RefResolver } from "../src/handler/refResolver";

interface CompletionTestCase {
    path: string;
    offset: number;
}

beforeAll(() => {
    App.init(path.join(getDebugDir(), 'storage'));
});

beforeEach(async () => {
    await App.clearCache();
});

afterAll(async() => {
    await App.shutdown();
});

describe('completion', () => {
    it('query from RefResolver', async () => {
        const indexer = App.get<Indexer>(Indexer);
        const refTable = App.get<ReferenceTable>(ReferenceTable);
        const phpDocTable = App.get<PhpDocumentTable>(PhpDocumentTable);

        const testCases: CompletionTestCase[] = [
            { path: path.join(getCaseDir(), 'completion', 'function.php'), offset: 18 },
            { path: path.join(getCaseDir(), 'completion', 'constant.php'), offset: 14 },
            { path: path.join(getCaseDir(), 'completion', 'constant.php'), offset: 29 },
            { path: path.join(getCaseDir(), 'completion', 'constant.php'), offset: 37 },
        ];

        await indexer.syncFileSystem(
            await PhpFileInfo.createFileInfo(path.join(getCaseDir(), 'global_symbols.php'))
        );
        await indexer.syncFileSystem(
            await PhpFileInfo.createFileInfo(path.join(getCaseDir(), 'function_declare.php'))
        );
        for (let testCase of testCases) {
            await indexer.syncFileSystem(await PhpFileInfo.createFileInfo(testCase.path));

            let uri = pathToUri(testCase.path);
            let phpDoc = await phpDocTable.get(uri);
            let ref = await refTable.findAt(uri, testCase.offset);

            if (phpDoc === null || ref === null) {
                continue;
            }

            let symbols = await RefResolver.searchSymbolsForReference(phpDoc, ref);
            for (let i = 0; i < symbols.length; i++) {
                symbols[i] = symbols[i].toObject();
            }

            expect(symbols).toMatchSnapshot();
        }
    });
});