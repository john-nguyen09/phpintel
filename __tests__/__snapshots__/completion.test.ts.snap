// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`completion completion for constant access references 1`] = `
Array [
  Function {
    "description": "Officia aliquip adipisicing et nulla et laboris dolore labore.
",
    "name": TypeName {
      "name": "\\\\test_function",
    },
    "parameters": Array [],
    "scopeVar": ScopeVar {
      "variables": Object {},
    },
    "types": Array [],
  },
  Function {
    "description": "",
    "name": TypeName {
      "name": "\\\\test_function1",
    },
    "parameters": Array [
      Parameter {
        "location": Location {},
        "name": "$stringParam",
        "refKind": 3,
        "scope": null,
        "type": TypeComposite {
          "_types": Array [
            TypeName {
              "name": "string",
            },
          ],
        },
        "value": "",
      },
      Parameter {
        "location": Location {},
        "name": "$intParam",
        "refKind": 3,
        "scope": null,
        "type": TypeComposite {
          "_types": Array [
            TypeName {
              "name": "int",
            },
          ],
        },
        "value": "",
      },
      Parameter {
        "location": Location {},
        "name": "$boolParam",
        "refKind": 3,
        "scope": null,
        "type": TypeComposite {
          "_types": Array [
            TypeName {
              "name": "boolean",
            },
          ],
        },
        "value": "",
      },
      Parameter {
        "location": Location {},
        "name": "$floatParam",
        "refKind": 3,
        "scope": null,
        "type": TypeComposite {
          "_types": Array [
            TypeName {
              "name": "float",
            },
          ],
        },
        "value": "",
      },
    ],
    "scopeVar": ScopeVar {
      "variables": Object {
        "$boolParam": TypeComposite {
          "_types": Array [
            TypeName {
              "name": "boolean",
            },
          ],
        },
        "$floatParam": TypeComposite {
          "_types": Array [
            TypeName {
              "name": "float",
            },
          ],
        },
        "$intParam": TypeComposite {
          "_types": Array [
            TypeName {
              "name": "int",
            },
          ],
        },
        "$stringParam": TypeComposite {
          "_types": Array [
            TypeName {
              "name": "string",
            },
          ],
        },
        "$test": TypeComposite {
          "_types": Array [
            TypeName {
              "name": "int",
            },
          ],
        },
      },
    },
    "types": Array [
      TypeName {
        "name": "boolean",
      },
      TypeName {
        "name": "string",
      },
      TypeName {
        "name": "int",
      },
      TypeName {
        "name": "float",
      },
      TypeName {
        "name": "\\\\TestClass",
      },
    ],
  },
  Function {
    "description": "Test function 2",
    "name": TypeName {
      "name": "\\\\test_function2",
    },
    "parameters": Array [
      Parameter {
        "location": Location {},
        "name": "$param1",
        "refKind": 3,
        "scope": null,
        "type": TypeComposite {
          "_types": Array [
            TypeName {
              "name": "\\\\array|int",
            },
          ],
        },
        "value": "",
      },
      Parameter {
        "location": Location {},
        "name": "$param2",
        "refKind": 3,
        "scope": null,
        "type": TypeComposite {
          "_types": Array [
            TypeName {
              "name": "string",
            },
          ],
        },
        "value": "",
      },
    ],
    "scopeVar": ScopeVar {
      "variables": Object {
        "$param1": TypeComposite {
          "_types": Array [
            TypeName {
              "name": "\\\\array|int",
            },
          ],
        },
        "$param2": TypeComposite {
          "_types": Array [
            TypeName {
              "name": "string",
            },
          ],
        },
      },
    },
    "types": Array [],
  },
]
`;

exports[`completion completion for constant access references 2`] = `
Array [
  Constant {
    "name": TypeName {
      "name": "\\\\TEST_CONST",
    },
    "type": TypeName {
      "name": "int",
    },
    "value": "123",
  },
  Constant {
    "name": TypeName {
      "name": "\\\\TEST_CONST1",
    },
    "type": TypeName {
      "name": "int",
    },
    "value": "456",
  },
  Constant {
    "name": TypeName {
      "name": "\\\\TEST_CONST2",
    },
    "type": TypeName {
      "name": "int",
    },
    "value": "789",
  },
  Constant {
    "name": TypeName {
      "name": "\\\\TEST_CONST_STRING",
    },
    "type": TypeName {
      "name": "string",
    },
    "value": "'Simple string'",
  },
  Constant {
    "name": TypeName {
      "name": "\\\\TEST_CONST_STRING2",
    },
    "type": TypeName {
      "name": "string",
    },
    "value": "'Not so simple string'",
  },
  Constant {
    "name": TypeName {
      "name": "\\\\TEST_CONST_STRING3",
    },
    "type": TypeName {
      "name": "string",
    },
    "value": "'Quite complex ' . TEST_CONST_STRING",
  },
  DefineConstant {
    "name": TypeName {
      "name": "\\\\TEST_DEFINE_CONSTANT",
    },
    "type": TypeName {
      "name": "boolean",
    },
    "value": "false",
  },
  DefineConstant {
    "name": TypeName {
      "name": "\\\\TEST_STRING_CONSTANT1",
    },
    "type": TypeName {
      "name": "string",
    },
    "value": "'simple string'",
  },
  DefineConstant {
    "name": TypeName {
      "name": "\\\\TEST_STRING_CONSTANT2",
    },
    "type": TypeName {
      "name": "string",
    },
    "value": "'not so simple string'",
  },
  DefineConstant {
    "name": TypeName {
      "name": "\\\\TEST_STRING_CONSTANT3",
    },
    "type": TypeName {
      "name": "string",
    },
    "value": "'more of ' . TEST_STRING_CONSTANT2",
  },
]
`;

exports[`completion completion for constant access references 3`] = `
Array [
  Constant {
    "name": TypeName {
      "name": "\\\\TEST_CONST_STRING",
    },
    "type": TypeName {
      "name": "string",
    },
    "value": "'Simple string'",
  },
  Constant {
    "name": TypeName {
      "name": "\\\\TEST_CONST_STRING2",
    },
    "type": TypeName {
      "name": "string",
    },
    "value": "'Not so simple string'",
  },
  Constant {
    "name": TypeName {
      "name": "\\\\TEST_CONST_STRING3",
    },
    "type": TypeName {
      "name": "string",
    },
    "value": "'Quite complex ' . TEST_CONST_STRING",
  },
]
`;

exports[`completion completion for constant access references 4`] = `
Array [
  Constant {
    "name": TypeName {
      "name": "\\\\TEST_CONST",
    },
    "type": TypeName {
      "name": "int",
    },
    "value": "123",
  },
  Constant {
    "name": TypeName {
      "name": "\\\\TEST_CONST1",
    },
    "type": TypeName {
      "name": "int",
    },
    "value": "456",
  },
  Constant {
    "name": TypeName {
      "name": "\\\\TEST_CONST2",
    },
    "type": TypeName {
      "name": "int",
    },
    "value": "789",
  },
  DefineConstant {
    "name": TypeName {
      "name": "\\\\TEST_DEFINE_CONSTANT",
    },
    "type": TypeName {
      "name": "boolean",
    },
    "value": "false",
  },
  DefineConstant {
    "name": TypeName {
      "name": "\\\\TEST_STRING_CONSTANT1",
    },
    "type": TypeName {
      "name": "string",
    },
    "value": "'simple string'",
  },
  DefineConstant {
    "name": TypeName {
      "name": "\\\\TEST_STRING_CONSTANT2",
    },
    "type": TypeName {
      "name": "string",
    },
    "value": "'not so simple string'",
  },
  DefineConstant {
    "name": TypeName {
      "name": "\\\\TEST_STRING_CONSTANT3",
    },
    "type": TypeName {
      "name": "string",
    },
    "value": "'more of ' . TEST_STRING_CONSTANT2",
  },
  DefineConstant {
    "name": TypeName {
      "name": "\\\\CONSTANT_SAMPLE1",
    },
    "type": TypeName {
      "name": "boolean",
    },
    "value": "false",
  },
  Constant {
    "name": TypeName {
      "name": "\\\\TEST_CONST_STRING",
    },
    "type": TypeName {
      "name": "string",
    },
    "value": "'Simple string'",
  },
  Constant {
    "name": TypeName {
      "name": "\\\\TEST_CONST_STRING2",
    },
    "type": TypeName {
      "name": "string",
    },
    "value": "'Not so simple string'",
  },
  Constant {
    "name": TypeName {
      "name": "\\\\TEST_CONST_STRING3",
    },
    "type": TypeName {
      "name": "string",
    },
    "value": "'Quite complex ' . TEST_CONST_STRING",
  },
]
`;

exports[`completion completion for current scoped member references 1`] = `
Array [
  Property {
    "description": "",
    "location": Location {
      "range": Range {
        "end": 106,
        "start": 93,
      },
      "relativeUri": "case/completion/currentScopedMember.php",
    },
    "modifier": SymbolModifier {
      "modifier": 32,
      "visibility": 2,
    },
    "name": "$NUM_CPUS",
    "scope": TypeName {
      "name": "\\\\CurrentScopedMember",
    },
  },
]
`;

exports[`completion completion for current scoped member references 2`] = `
Array [
  Method {
    "description": "",
    "location": Location {
      "range": Range {
        "end": 2290,
        "start": 2236,
      },
      "relativeUri": "case/global_symbols.php",
    },
    "modifier": SymbolModifier {
      "modifier": 32,
      "visibility": 1,
    },
    "name": TypeName {
      "name": "staticTestMethod",
    },
    "scope": null,
  },
  Property {
    "description": "",
    "location": Location {
      "range": Range {
        "end": 2203,
        "start": 2184,
      },
      "relativeUri": "case/global_symbols.php",
    },
    "modifier": SymbolModifier {
      "modifier": 32,
      "visibility": 1,
    },
    "name": "$staticTestProperty",
    "scope": null,
  },
  ClassConstant {
    "location": Location {
      "range": Range {
        "end": 2164,
        "start": 2142,
      },
      "relativeUri": "case/global_symbols.php",
    },
    "name": TypeName {
      "name": "TEST_CLASS_CONST",
    },
    "scope": null,
  },
]
`;

exports[`completion completion for current scoped member references 3`] = `
Array [
  Method {
    "description": "",
    "location": Location {
      "range": Range {
        "end": 184,
        "start": 115,
      },
      "relativeUri": "case/completion/currentScopedMember.php",
    },
    "modifier": SymbolModifier {
      "modifier": 32,
      "visibility": 1,
    },
    "name": TypeName {
      "name": "GetNumCpus",
    },
    "scope": TypeName {
      "name": "\\\\CurrentScopedMember",
    },
  },
  Property {
    "description": "",
    "location": Location {
      "range": Range {
        "end": 106,
        "start": 93,
      },
      "relativeUri": "case/completion/currentScopedMember.php",
    },
    "modifier": SymbolModifier {
      "modifier": 32,
      "visibility": 2,
    },
    "name": "$NUM_CPUS",
    "scope": TypeName {
      "name": "\\\\CurrentScopedMember",
    },
  },
  ClassConstant {
    "location": Location {
      "range": Range {
        "end": 67,
        "start": 48,
      },
      "relativeUri": "case/completion/currentScopedMember.php",
    },
    "name": TypeName {
      "name": "SYSTEM_NUM_CPUS",
    },
    "scope": TypeName {
      "name": "\\\\CurrentScopedMember",
    },
  },
]
`;

exports[`completion completion for scoped member references 1`] = `
Array [
  Method {
    "description": "Do magna consequat veniam minim proident eiusmod incididunt aute proident.",
    "location": Location {
      "range": Range {
        "end": 1583,
        "start": 1493,
      },
      "relativeUri": "case/global_symbols.php",
    },
    "modifier": SymbolModifier {
      "modifier": 32,
      "visibility": 1,
    },
    "name": TypeName {
      "name": "staticTestMethod",
    },
    "scope": TypeName {
      "name": "\\\\TestClass",
    },
  },
  Property {
    "description": "Lorem excepteur officia sit anim velit veniam enim.
",
    "location": Location {
      "range": Range {
        "end": 1266,
        "start": 1247,
      },
      "relativeUri": "case/global_symbols.php",
    },
    "modifier": SymbolModifier {
      "modifier": 32,
      "visibility": 1,
    },
    "name": "$staticTestProperty",
    "scope": TypeName {
      "name": "\\\\TestClass",
    },
  },
  ClassConstant {
    "location": Location {
      "range": Range {
        "end": 1120,
        "start": 1098,
      },
      "relativeUri": "case/global_symbols.php",
    },
    "name": TypeName {
      "name": "TEST_CLASS_CONST",
    },
    "scope": TypeName {
      "name": "\\\\TestClass",
    },
  },
]
`;

exports[`completion completion for scoped member references 2`] = `
Array [
  Property {
    "description": "Lorem excepteur officia sit anim velit veniam enim.
",
    "location": Location {
      "range": Range {
        "end": 1266,
        "start": 1247,
      },
      "relativeUri": "case/global_symbols.php",
    },
    "modifier": SymbolModifier {
      "modifier": 32,
      "visibility": 1,
    },
    "name": "$staticTestProperty",
    "scope": TypeName {
      "name": "\\\\TestClass",
    },
  },
]
`;

exports[`completion completion for scoped member references 3`] = `
Array [
  Method {
    "description": "Do magna consequat veniam minim proident eiusmod incididunt aute proident.",
    "location": Location {
      "range": Range {
        "end": 1583,
        "start": 1493,
      },
      "relativeUri": "case/global_symbols.php",
    },
    "modifier": SymbolModifier {
      "modifier": 32,
      "visibility": 1,
    },
    "name": TypeName {
      "name": "staticTestMethod",
    },
    "scope": TypeName {
      "name": "\\\\TestClass",
    },
  },
  ClassConstant {
    "location": Location {
      "range": Range {
        "end": 1120,
        "start": 1098,
      },
      "relativeUri": "case/global_symbols.php",
    },
    "name": TypeName {
      "name": "TEST_CLASS_CONST",
    },
    "scope": TypeName {
      "name": "\\\\TestClass",
    },
  },
]
`;

exports[`completion completion for scoped member references 4`] = `
Array [
  ClassConstant {
    "location": Location {
      "range": Range {
        "end": 1120,
        "start": 1098,
      },
      "relativeUri": "case/global_symbols.php",
    },
    "name": TypeName {
      "name": "TEST_CLASS_CONST",
    },
    "scope": TypeName {
      "name": "\\\\TestClass",
    },
  },
]
`;

exports[`completion completion for scoped member references 5`] = `Array []`;
